plugins/wasmedge_rwkv/rwkv.cpp

```diff
@@ -0,0 +1,258 @@
+
+#include "rwkv_error_handling.inc"
+
+#include "rwkv_utilities.inc"
+
+#include "rwkv_file_format.inc"
+
+#include "rwkv_model_loading.inc"
+
+#include "rwkv_operators.inc"
+
+#include "rwkv_graph.inc"
+
+// API function.
+struct rwkv_context * rwkv_init_from_file(const char * file_path, const uint32_t n_threads, const uint32_t n_gpu_layers) {
+    global_last_error = RWKV_ERROR_NONE;
+
+    std::unique_ptr<struct rwkv_context> ctx(new(std::nothrow) struct rwkv_context());
+    RWKV_ASSERT_NULL_MSG(RWKV_ERROR_CTX | RWKV_ERROR_ALLOC, ctx, "Failed to allocate rwkv_context");
+
+    ctx->model = new(std::nothrow) struct rwkv_model();
+    ctx->model->reference_count++;
+
+    ctx->n_threads = n_threads;
+
+    if (n_gpu_layers) {
+        ggml_backend_t backend = nullptr;
+
+#ifdef GGML_USE_CUDA
+        backend = ggml_backend_cuda_init(0);
+        RWKV_ENSURE_OR_NULL(backend);
+#endif
+
+#ifdef GGML_USE_METAL
+        backend = ggml_backend_metal_init();
+        RWKV_ENSURE_OR_NULL(backend);
+#endif
+
+#ifdef GGML_USE_BLAS
+        backend = ggml_backend_blas_init();
+        RWKV_ENSURE_OR_NULL(backend);
+        ggml_backend_blas_set_n_threads(backend, ctx->n_threads);
+#endif
+        if (backend != nullptr) {
+            ctx->model->backends.push_back(backend);
+        }
+    }
+
+    ggml_backend_t cpu_backend = ggml_backend_cpu_init();
+    RWKV_ENSURE_OR_NULL(cpu_backend);
+    ggml_backend_cpu_set_n_threads(cpu_backend, n_threads);
+    ctx->model->backends.push_back(cpu_backend);
+
+    int ngl = n_gpu_layers;
+    if (ctx->model->backends.size() == 1) {
+        ngl = 0;
+    }
+
+    RWKV_ENSURE_OR_NULL(rwkv_load_model_from_file(file_path, *ctx->model, ngl));
+
+    RWKV_ENSURE_OR_NULL(rwkv_measure_and_build_serial_context(*ctx->model, ctx->serial_graph));
+
+    return ctx.release();
+}
+
+// API function.
+struct rwkv_context * rwkv_clone_context(struct rwkv_context * ctx, const uint32_t n_threads) {
+    std::unique_ptr<struct rwkv_context> clone(new(std::nothrow) struct rwkv_context());
+    RWKV_ASSERT_NULL_MSG(RWKV_ERROR_CTX | RWKV_ERROR_ALLOC, clone, "Failed to allocate rwkv_context");
+
+    clone->model = ctx->model;
+    clone->model->reference_count++;
+
+    clone->n_threads = n_threads;
+
+    RWKV_ENSURE_OR_NULL(rwkv_measure_and_build_serial_context(*clone->model, clone->serial_graph));
+
+    clone->last_used_sequence_length = 0;
+
+    clone->print_errors = ctx->print_errors;
+
+    return clone.release();
+}
+
+#include "rwkv_eval.inc"
+
+// API function.
+// Provided for backwards compatibility.
+extern "C" RWKV_API uint32_t rwkv_get_state_buffer_element_count(const struct rwkv_context * ctx) {
+    return rwkv_get_state_len(ctx);
+}
+
+// API function.
+// Provided for backwards compatibility.
+extern "C" RWKV_API uint32_t rwkv_get_logits_buffer_element_count(const struct rwkv_context * ctx) {
+    return rwkv_get_logits_len(ctx);
+}
+
+// API function.
+size_t rwkv_get_n_vocab(const struct rwkv_context * ctx) {
```

Comment:
```suggestion
size_t rwkv_get_n_vocab(const struct rwkv_context * ctx) {
if (!ctx || !ctx->model) {
    return 0;
}
```

we should probably add null checks for all getter functions here